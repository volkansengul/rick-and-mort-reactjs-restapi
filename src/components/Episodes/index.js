import React from 'react';
import Api from '../../api';

export default class Index extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            episodeIdList: props.episodeIdList,
            episodes: null
        };
        this.api = new Api();
    }

    getEpisodeIdList() {
        if (this.state.episodeIdList !== null) {
            return this.state.episodeIdList.map((item) => {
                return item.replace('https://rickandmortyapi.com/api/episode/', '')
            });
        } else {
            return false;
        }
    }

    getEpisodes() {
        if (this.getEpisodeIdList()) {
            this.api.episodes(this.getEpisodeIdList()).then(resp => {
                let episodeList = Array.isArray(resp) ? resp.sort()
                    .reverse()
                    .slice(0, (this.props.limit ? this.props.limit : 5)) : [resp];
                const episodes = episodeList.map((item, i) => {
                    return (
                        <div key={'episode_' + i} className={'episode'}>
                            <h3>{item.name} <i>({item.episode})</i></h3>
                            <span className={'episode-air-date'}>{item.air_date}</span>
                            <div className={'clear'}/>
                        </div>
                    )
                });
                this.setState({
                    episodes: episodes
                });
            })
        }
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        if (nextProps.episodeIdList !== null) {
            return {episodeIdList: nextProps.episodeIdList};
        } else {
            return null;
        }
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevState.episodeIdList !== this.state.episodeIdList) {
            this.getEpisodes();
        }
    }


    render() {
        return (
            <div className={'episodes'}>
                <h3 className={'title'}>Last 5 Episodes</h3>
                {this.state.episodes}
            </div>
        )
    }
}