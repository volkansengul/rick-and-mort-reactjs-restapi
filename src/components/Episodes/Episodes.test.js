import React from 'react';
import {BrowserRouter as Router} from 'react-router-dom';
import renderer from 'react-test-renderer';

import Episodes from './index';

const episodes = [
    {
        name: 'Rest and Ricklaxation',
        episode: '(S03E06)',
        air_date: 'August 27, 2017'
    },
    {
        name: 'The Ricklantis Mixup',
        episode: '(S03E07)',
        air_date: 'September 10, 2017'
    }
];

it('Character Episode List Test', () => {
    const episodeList = renderer.create(
        <Router>
            <Episodes episodeIdList={episodes} limit={2}/>
        </Router>
    ).toJSON();
    expect(episodeList).toMatchSnapshot();
});
