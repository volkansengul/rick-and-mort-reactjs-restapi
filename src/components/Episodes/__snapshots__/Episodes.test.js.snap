// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Episode List Test 1`] = `
<div
  className="episodes"
>
  <h3
    className="title"
  >
    Last 5 Episodes
  </h3>
</div>
`;
